src/brevitas_examples/llm/llm_quant/learned_round_utils.py

```diff
@@ -0,0 +1,213 @@
+
+    def init_cache(self) -> Any:
+        # cache_args, cache_kwargs, cache_outs
+        return [], [], []
+
+    def populate_cache(
+        self,
+        cache: Any,
+        model: nn.Module,
+        block: nn.Module,
+        data_loader: DataLoader,
+        keep_gpu: bool = True,
+        **kwargs,
+    ) -> int:
+        # Unpack cache
+        cache_args, cache_kwargs, cache_outs = cache
+        # Cache needs to be cleaned between blocks. No need to clear the
+        # kwargs cache, as this is only updated for the first block.
+        cache_args.clear()
+        cache_outs.clear()
+        # Save FP output
+        self._save_inp_out_data(
+            model,
+            block,
+            data_loader,
+            cache_args,
+            cache_kwargs,
+            cache_outs,
+            store_args=False,
+            store_kwargs=False,
+            store_outs=True,
+            keep_gpu=keep_gpu,
+            disable_quant=True)
+        # Save Quant input
+        self._save_inp_out_data(
+            model,
+            block,
+            data_loader,
+            cache_args,
+            cache_kwargs,
+            cache_outs,
+            store_args=True,
+            store_kwargs=len(cache_kwargs) == 0,
+            store_outs=False,
+            keep_gpu=keep_gpu,
+            disable_quant=False)
+        # Return number of samples in calibration set
+        return len(cache_args)
+
+    def sample_cache(
+        self,
+        block: nn.Module,
+        cache: Any,
+        indices: torch.Tensor,
+        input_dim: int = 0,
+        **kwargs_fn,
+    ) -> Tuple[Any, torch.Tensor]:
+        cache_args, cache_kwargs, cache_outs = cache
+        device = next(block.parameters()).device
+        # Positional arguments
+        args = [cache_args[i] for i in indices]
+        args = tuple(torch.cat(arg_tensor, dim=0) for arg_tensor in zip(*args))
+        # Keyword arguments
+        kwargs_dict = [cache_kwargs[i] for i in indices]
+        kwargs = {}
+        for curr_dict in kwargs_dict:
+            for key, value in curr_dict.items():
+                if isinstance(value, torch.Tensor):
+                    if key not in kwargs:
+                        kwargs[key] = []
+                    kwargs[key].append(value)
+                else:
+                    if key not in kwargs:
+                        kwargs[key] = value
+        for key, value in kwargs.items():
+            if isinstance(value, list) and len(value) > 0:
+                kwargs[key] = torch.cat(kwargs[key], dim=input_dim)
+        # FP outputs
+        outs = torch.cat([cache_outs[i] for i in indices], dim=input_dim)
+        # Make sure that the inputs and outputs are in the same device as block,
+        # before running its forward pass.
+        args = send_to_device(args, device)
+        kwargs = send_to_device(kwargs, device)
+        outs = send_to_device(outs, device)
+
+        return (args, kwargs), outs
+
+    def run_forward(
```

Comment:
This shouldn't be here. Rather, the user should be able to specify whatever interface for the input to the model.
We should provide the interface to accept a function with a certain signature, and the user decides what happens inside that function.

Possible signature
```python
def model_forward(model, model_args, model_kwargs):
```

Reply:
This signature is defined in LearnedRoundModelUtils.